### ISSUE TITLE
h5 float file to cool: float converted to integer?

### ISSUE THREAD

**Issue by gtrichard** [deeptools/HiCExplorer#289]
I did

```
hicConvertFormat  --matrices matrix.h5 --outFileName matrix.cool --inputFormat h5 --outputFormat cool
```

It worked like a charm, but when I 

```
cooler dump matrix.cool
```

I receive this output:

```
11	11	23
11	12	99
11	13	74
11	14	44
11	15	28
11	16	32
11	17	26
11	18	10
11	19	9
11	20	17
11	21	5
11	22	13
11	23	12
11	25	9
11	26	8
11	27	6
11	28	6
11	29	3
11	30	5
```

So the counts are integers while my h5 input was a float (I verified it using the GInteractions format, btw the GInteractions format as output is not working on hicConvertFormat but works on hicExport...). I think it ressembles that issue https://github.com/mirnylab/cooler/issues/129

Is the conversion from h5 to cool is checking for int or float input and use the `--count-as-float` option from cooler?

**Comment by gtrichard** [comment]
After more digging, is it intended to increase the compatibility of the .cool file with other softwares?

**Comment by joachimwolff** [comment]
I would be happy to support the 'standard' cooler but to my experience many things do miss for this:

- The standard for 'counts' is integer, but you can change it to float if you like (That was actually the answer when I ask how to deal with this)
- I think it is non-sense to enforce integer, float make way more sense after correction (and side note: It is restricted to integer but negative inegers are allowed...)
- The correction factors should be stored in 'weights' but e.g. hic2cool doesn't care here and uses three new columns (KV, VC and VC_SQRT)
- To apply the correction every time on raw data is resulting in inaccurate results (rounding errors etc. see hicAggregateContacts, use it with cool and h5, it gives you different results)
- rounding float to integer is inaccurate
- there is no clear way how to store multiple matrices in one. Should it be, as `hic2cool` is doing it, stored in `:/resolutions/10000`? Or as others do `:/1`? How to combine different samples? And how do tools like HiGlass deal with this?
- Some use division to apply the correction (aka HiCExplorer), some use a multiplication.

So you see there are many things that are not specified or clear and one tool is implementing it in this way, another one somehow else. Some of these issues "Should be discussed" as cooler people said. Maybe they did it in the last half year or don't, I did not really followed it. I am just a bit annoyed by them, for me it does not seem they thought this file format really through.

**Comment by gtrichard** [comment]
Oh I get it, it is much more complicated than I thought... But it also explaining very well my feelings since I started analysing Hi-C data: nothing is standard depending on which software you use, which makes it really hard to get the bottomline or "rule of thumb" for this and that in Hi-C data analysis... and which also makes it really to go from one tool to another. Since also they all use different normalisation methods as you pointed out and as we saw for hicPCA recently, it makes it really hard to create standard procedures and file formats.

About hic2cool and the weights, right now for hicex to cool, I first convert h5 to GInteractions (named bg2 by cooler people) and use this command line to convert the file to .cool using cooler:

```
cooler cload pairs --assembly dm6 --tril-action drop --zero-based dmel.chrom.size:529 matrix.bg2.gz matrix.529.cool -c1 1 -p1 2 -c2 4 -p2 5
```

Here 529 is the median bin size given by hicInfo (as I was dealing with RF resolution matrix that I had to convert to fixed bins for HiGlass). That way I'm avoiding to use hic2cool, perhaps it can help you? And it also deals with RF resolution matrices to fixed bins size (but the issue is also that you lose the RF resolution information).

For hicAggregateContacts, I don't understand how taking a corrected matrix in h5, and convert it to cool change the values stored (outside of the float to integer transformation) ?

As I see it, cooler is officially supporting float numbers, now the question is, are other tools (like HiGlass) using cooler format support floating numbers? But anyway it is true that the cool format is still lacking a proper delineation (as you mentioned for the difference between `hic2cool` and `cooler zoomify` for the mcool format) and is still evolving...
